crates/apollo-router-core/src/query_planner/caching_query_planner.rs

```diff
@@ -36,26 +38,73 @@ impl<T: QueryPlanner> QueryPlanner for CachingQueryPlanner<T> {
-            .cached
-            .lock()
-            .await
-            .get(&(query.clone(), operation.clone(), options.clone()))
-            .cloned()
-        {
+    ) -> PlanResult {
+        let mut locked_cache = self.cached.lock().await;
+        let key = (query, operation, options);
+        if let Some(value) = locked_cache.get(&key).cloned() {
             return value;
         }
 
-        let value = self
-            .delegate
-            .get(query.clone(), operation.clone(), options.clone())
-            .await;
-        self.cached
-            .lock()
-            .await
-            .insert((query, operation, options), value.clone());
-        value
+        // Holding a lock across the delegated get is a bad idea because
+        // the delegate get() calls into v8 for processing of the plan.
+        // This would block all other get() requests for a potentially
+        // long time.
+        // Alternatively, if we don't hold the lock, there is a risk
+        // that we will do the work multiple times. This is also
+        // sub-optimal.
+
+        // To work around this, we keep a list of keys we are currently
+        // processing in the delegate. If we try to get a key on this
+        // list, we block and wait for it to complete and then retry.
+        //
+        // This is more complex than either of the two simple
+        // alternatives but succeeds in providing a mechanism where each
+        // client only waits for uncached QueryPlans they are going to
+        // use AND avoids generating the plan multiple times.
+
+        drop(locked_cache);
+
+        let mut locked_wait_map = self.wait_map.lock().await;
+
+        match locked_wait_map.get_mut(&key) {
+            Some(waiter) => {
+                // Register interest in key
+                let mut receiver = waiter.subscribe();
+                drop(locked_wait_map);
+                let msg = receiver.recv().await?;
+                assert_eq!(msg.0, key);
+                msg.1
+            }
+            None => {
+                let (tx, _rx) = broadcast::channel(10);
+                locked_wait_map.insert(key.clone(), tx.clone());
+                drop(locked_wait_map);
+                // This is the potentially high duration operation
+                let value = self
+                    .delegate
+                    .get(key.0.clone(), key.1.clone(), key.2.clone())
+                    .await;
+                // Update our cache
+                let mut locked_cache = self.cached.lock().await;
+                locked_cache.put(key.clone(), value.clone());
+                // Update our wait list
+                let mut locked_wait_map = self.wait_map.lock().await;
+                locked_wait_map.remove(&key);
+                // Let our waiters know
+                let broadcast_value = value.clone();
+                tokio::task::spawn_blocking(move || tx.send((key, broadcast_value))).await??;
```

Comment:
According to the [documentation](https://docs.rs/tokio/1.14.0/tokio/sync/broadcast/struct.Sender.html#method.send):

> Attempts to send a value to all active Receiver handles, returning it back if it could not be sent.
> 
> A successful send occurs when there is at least one active Receiver handle. An unsuccessful send would be one where all associated Receiver handles have already been dropped.

It is possible that the receiver have been canceled for some reason. But this is no reason to fail the current request. I think you can just safely ignore this error:

```rust
let _ = tx.send((key, broadcast_value))
```

Again, just like RecvError, we don't want to add the SendError to the QueryPlannerError as this can be returned to the user making the call. So either we make a strategy to "retry" to prevent failing or we unwrap because we are 100% confident that this will not fail. Correct me if I'm wrong but I think here we can just ignore any error from sending the message.

Reply:
(Repeat from above) I agree that the documentation suggest this is safe to do, so I've done it.